src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -62,27 +66,51 @@ pub struct LoanDataFlowOperator;
 
 pub type LoanDataFlow<'a, 'tcx> = DataFlowContext<'a, 'tcx, LoanDataFlowOperator>;
 
-pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>) {
-    for body_owner_def_id in tcx.body_owners() {
-        tcx.borrowck(body_owner_def_id);
+pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, save_analysis: bool) -> Option<HashMap<DefId, AnalysisResult<'a, 'tcx>>> {
+    // FIXME: nashenas88 support this through TyCtxt
+    if save_analysis {
+        let mut map = HashMap::new();
+        for body_owner_def_id in tcx.body_owners() {
+            let result = borrowck(tcx, body_owner_def_id, save_analysis).unwrap();
+            map.insert(body_owner_def_id, result);
+        }
+        Some(map)
+    } else {
+        for body_owner_def_id in tcx.body_owners() {
+            tcx.borrowck(body_owner_def_id);
+        }
+        None
     }
 }
 
 pub fn provide(providers: &mut Providers) {
     *providers = Providers {
-        borrowck,
+        borrowck: borrowck_provider,
         ..*providers
     };
 }
 
 /// Collection of conclusions determined via borrow checker analyses.
 pub struct AnalysisData<'a, 'tcx: 'a> {
+    pub safe_loans: Vec<SafeLoan>,
     pub all_loans: Vec<Loan<'tcx>>,
     pub loans: DataFlowContext<'a, 'tcx, LoanDataFlowOperator>,
     pub move_data: move_data::FlowedMoveData<'a, 'tcx>,
 }
 
-fn borrowck<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, owner_def_id: DefId) {
+pub struct AnalysisResult<'a, 'tcx: 'a> {
+    pub bccx: BorrowckCtxt<'a, 'tcx>,
```

Comment:
I'm not crazy about keeping the `BorrowckCtxt` around either. This feels to me like a "temporary" data structure, not one that is meant to be preserved. =) But I guess I'll have to read on to see what you're using it for exactly.